File: 01_language_comparison/dynamic_programming/knapsack.rb
class Knapsack
    attr_reader :weights, :values, :capacity
    
    def initialize(weights, values, capacity)
      @weights = weights
      @values = values
      @capacity = capacity
      @memo = {}
    end
    
    def solve
      n = @weights.length
      result = knapsack_recursive(n - 1, @capacity)
      
      # Reconstruct the solution (which items were selected)
      selected_items = []
      remaining_capacity = @capacity
      
      (n - 1).downto(0) do |i|
        if i == 0
          selected_items.unshift(0) if remaining_capacity >= @weights[0] && @values[0] > 0
        elsif knapsack_recursive(i, remaining_capacity) != knapsack_recursive(i - 1, remaining_capacity)
          selected_items.unshift(i)
          remaining_capacity -= @weights[i]
        end
      end
      
      return {
        max_value: result,
        selected_items: selected_items
      }
    end
    
    private
    
    def knapsack_recursive(i, w)
      # Base case
      return 0 if i < 0 || w <= 0
      
      # Check if already computed
      key = "#{i},#{w}"
      return @memo[key] if @memo.key?(key)
      
      # If weight of item is more than capacity, skip this item
      if @weights[i] > w
        result = knapsack_recursive(i - 1, w)
      else
        # Try taking the item and not taking the item, pick the max value
        take_item = @values[i] + knapsack_recursive(i - 1, w - @weights[i])
        skip_item = knapsack_recursive(i - 1, w)
        result = [take_item, skip_item].max
      end
      
      # Memoize the result
      @memo[key] = result
      return result
    end
  end
  
  def run_knapsack
    start_time = Time.now
    
    weights = [10, 20, 30]
    values = [60, 100, 120]
    capacity = 50
    
    knapsack = Knapsack.new(weights, values, capacity)
    result = knapsack.solve
    
    puts "Maximum value: #{result[:max_value]}"
    puts "Selected items: #{result[:selected_items].join(', ')}"
    
    end_time = Time.now
    puts "Execution time: #{(end_time - start_time) * 1000} ms"
  end
  
  run_knapsack

File: 01_language_comparison/dynamic_programming/knapsack.go
package main

import (
	"fmt"
	"time"
)

// KnapsackSolution holds the solution data
type KnapsackSolution struct {
	MaxValue     int
	SelectedItems []int
}

// SolveKnapsack solves the 0/1 knapsack problem
func SolveKnapsack(weights []int, values []int, capacity int) KnapsackSolution {
	n := len(weights)
	
	// Create a 2D array for dynamic programming
	dp := make([][]int, n+1)
	for i := range dp {
		dp[i] = make([]int, capacity+1)
	}
	
	// Fill the dp table
	for i := 0; i <= n; i++ {
		for w := 0; w <= capacity; w++ {
			if i == 0 || w == 0 {
				dp[i][w] = 0
			} else if weights[i-1] <= w {
				// Max of including or excluding current item
				dp[i][w] = max(values[i-1]+dp[i-1][w-weights[i-1]], dp[i-1][w])
			} else {
				dp[i][w] = dp[i-1][w]
			}
		}
	}
	
	// Backtrack to find selected items
	selectedItems := []int{}
	w := capacity
	for i := n; i > 0; i-- {
		if dp[i][w] != dp[i-1][w] {
			selectedItems = append([]int{i - 1}, selectedItems...) // Prepend to maintain order
			w -= weights[i-1]
		}
	}
	
	return KnapsackSolution{
		MaxValue:     dp[n][capacity],
		SelectedItems: selectedItems,
	}
}

// Helper function to get maximum of two integers
func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func main() {
	startTime := time.Now()
	
	weights := []int{10, 20, 30}
	values := []int{60, 100, 120}
	capacity := 50
	
	solution := SolveKnapsack(weights, values, capacity)
	
	fmt.Printf("Maximum value: %d\n", solution.MaxValue)
	fmt.Printf("Selected items: ")
	for i, item := range solution.SelectedItems {
		if i > 0 {
			fmt.Print(", ")
		}
		fmt.Print(item)
	}
	fmt.Println()
	
	executionTime := time.Since(startTime).Milliseconds()
	fmt.Printf("Execution time: %d ms\n", executionTime)
}